odl/contrib/solvers/spdhg/misc.py

```diff
@@ -0,0 +1,857 @@
+        return '{}({!r}, {!r})'.format(self.__class__.__name__, self.domain,
+                                       self.gamma)
+
+
+class KullbackLeiblerSmooth(odl.solvers.Functional):
+
+    """The smooth Kullback-Leibler divergence functional.
+
+    Notes
+    -----
+    If the functional is defined on an :math:`\mathbb{R}^n`-like space, the
+    smooth Kullback-Leibler functional :math:`\\phi` is defined as
+
+    .. math::
+        \\phi(x) = \\sum_{i=1}^n \\begin{cases}
+                x + r - y + y * \\log(y / (x + r))
+                    & \\text{if $x \geq 0$} \\
+                (y / (2 * r^2)) * x^2 + (1 - y / r) * x + r - b +
+                    b * \\log(b / r) & \\text{else}
+                                 \\end{cases}
+
+    where all variables on the right hand side of the equation have a subscript
+    i which is omitted for readability.
+
+    References
+    ----------
+    [CERS2017] Chambolle, A., Ehrhardt, M. J., Richtárik, P. and
+    Schönlieb, C.-B. *Stochastic Primal-Dual Hybrid Gradient Algorithm with
+    Arbitrary Sampling and Imaging Applications*.
+    ArXiv: http://arxiv.org/abs/1706.04957, 2017
+    """
+
+    def __init__(self, space, data, background):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        space : `DiscreteLp` or `FnBase`
+            Domain of the functional.
+        data : ``space`` `element-like`
+            Data vector which has to be non-negative.
+        background : ``space`` `element-like`
+            Background vector which has to be non-negative.
+        """
+
+        self.strong_convexity = 0
+
+        if background.ufuncs.less_equal(0).ufuncs.sum() > 0:
+            raise NotImplementedError('Background must be positive')
+
+        super().__init__(space=space, linear=False,
+                         grad_lipschitz=np.max(data/background**2))
+
+        if data not in self.domain:
+            raise ValueError('`data` not in `domain`'
+                             ''.format(data, self.domain))
+
+        self.__data = data
+        self.__background = background
+
+    @property
+    def data(self):
+        """The data in the Kullback-Leibler functional."""
+        return self.__data
+
+    @property
+    def background(self):
+        """The background in the Kullback-Leibler functional."""
+        return self.__background
+
+    def _call(self, x):
+        """Return the KL-diveregnce in the point ``x``.
+
+        If any components of ``x`` is non-positive, the value is positive
+        infinity.
+        """
+        y = self.data
+        r = self.background
+
+        # TODO: implement more efficiently in terms of memory and CPU/GPU
+        # TODO: cover properly the case y = 0
+
+        # x + r - y + y * log(y / (x + r)) = x - y * log(x + r) + c1
+        # with c1 = r - y + y * log y
+        i = x.ufuncs.greater_equal(0)
+        obj = (i * (x + r - y + y * (y / (x + r)).ufuncs.log()))
```

Comment:
I can already give you a heads-up here that this code will break after #1088 is merged. The reason is that `i` will be in a space with `dtype=bool`, and the subsequent operations will be illegal without cast to float. Boolean indexing will work, though.

Reply:
Will #1088 be also working on the GPU? It would be good if one would not need to distinguish the two cases.

Reply:
Not with odlcuda, but there are a bunch of other backends almost ready to go in. At least [cupy](https://github.com/cupy/cupy) does support everything we need.

Reply:
That sounds really good. I guess it is best for me to hang tight and leave the code the way it is. Once we have a good option we should change it accordingly.

Reply:
Yes, the plan is to push in #1088 first and then add a GPU backend (sorry for the intermediate breakage, but otherwise the PR never goes in).

Reply:
Is this now broken?

Reply:
It seems so:

Traceback (most recent call last):

  File "<ipython-input-1-6b456779813b>", line 131, in <module>
    callback=callback)

  File "/home/ehrhardt/repositories/github_myodl/odl/solvers/nonsmooth/primal_dual_hybrid_gradient.py", line 274, in pdhg
    proximal_dual_sigma(dual_tmp, out=y)

  File "/home/ehrhardt/repositories/github_myodl/odl/operator/operator.py", line 678, in __call__
    result = self._call_in_place(x, out=out, **kwargs)

  File "/home/ehrhardt/repositories/github_myodl/odl/operator/pspace_ops.py", line 229, in _call
    op(x[j], out=out[i])

  File "/home/ehrhardt/repositories/github_myodl/odl/operator/operator.py", line 678, in __call__
    result = self._call_in_place(x, out=out, **kwargs)

  File "/home/ehrhardt/repositories/github_myodl/odl/operator/operator.py", line 1364, in _call
    return self.left(tmp, out=out)

  File "/home/ehrhardt/repositories/github_myodl/odl/operator/operator.py", line 678, in __call__
    result = self._call_in_place(x, out=out, **kwargs)

  File "/home/ehrhardt/repositories/github_myodl/odl/operator/operator.py", line 1363, in _call
    self.right(x, out=tmp)

  File "/home/ehrhardt/repositories/github_myodl/odl/operator/operator.py", line 678, in __call__
    result = self._call_in_place(x, out=out, **kwargs)

  File "/home/ehrhardt/repositories/github_myodl/odl/contrib/solvers/spdhg/misc.py", line 842, in _call
    4 * s * y).ufuncs.sqrt()))

TypeError: unsupported operand type(s) for *: 'DiscreteLpElement' and 'DiscreteLpElement'

Reply:
Interestingly, I also can't cast it manually:

  File "<ipython-input-8-8a35290f7338>", line 1, in <module>
    x_opt = x_opt.astype('float')

  File "/home/ehrhardt/repositories/github_myodl/odl/discr/discretization.py", line 361, in astype
    return self.space.astype(dtype).element(self.tensor.astype(dtype))

  File "/home/ehrhardt/repositories/github_myodl/odl/space/base_tensors.py", line 259, in astype
    return self._astype(dtype)

  File "/home/ehrhardt/repositories/github_myodl/odl/discr/lp_discr.py", line 375, in _astype
    fspace = self.fspace.astype(dtype)

  File "/home/ehrhardt/repositories/github_myodl/odl/space/fspace.py", line 706, in astype
    if out_dtype == self.real_out_dtype:

  File "/home/ehrhardt/repositories/github_myodl/odl/space/fspace.py", line 282, in real_out_dtype
    raise TypeError('no real variant of output dtype defined')

TypeError: no real variant of output dtype defined

Reply:
That's an actual bug. See #1285

Reply:
Fixed now